packages/stats/src/main/c/model.c

```diff
@@ -0,0 +1,2005 @@
+
+    /* This section of the code checks the types of the variables
+       in the model frame.  Note that it should really only check
+       the variables if they appear in a term in the model.
+       Because it does not, we need to allow other types here, as they
+       might well occur on the LHS.
+       The R code converts all character variables in the model frame to
+       factors, so the only types that ought to be here are logical,
+       integer (including factor), numeric and complex.
+     */
+
+    PROTECT(variable = allocVector(VECSXP, nVar));
+    PROTECT(nlevs = allocVector(INTSXP, nVar));
+    PROTECT(ordered = allocVector(LGLSXP, nVar));
+    PROTECT(columns = allocVector(INTSXP, nVar));
+
+    for (i = 0; i < nVar; i++) {
+	var_i = SET_VECTOR_ELT(variable, i, VECTOR_ELT(vars, i));
+	if (nrows(var_i) != n)
+	    error(_("variable lengths differ (found for variable %d)"), i);
+	if (isOrdered_int(var_i)) {
+	    LOGICAL(ordered)[i] = 1;
+	    if((INTEGER(nlevs)[i] = nlevels(var_i)) < 1)
+		error(_("variable %d has no levels"), i+1);
+	    /* will get updated later when contrasts are set */
+	    INTEGER(columns)[i] = ncols(var_i);
+	}
+	else if (isUnordered_int(var_i)) {
+	    LOGICAL(ordered)[i] = 0;
+	    if((INTEGER(nlevs)[i] = nlevels(var_i)) < 1)
+		error(_("variable %d has no levels"), i+1);
+	    /* will get updated later when contrasts are set */
+	    INTEGER(columns)[i] = ncols(var_i);
+	}
+	else if (isLogical(var_i)) {
+	    LOGICAL(ordered)[i] = 0;
+	    INTEGER(nlevs)[i] = 2;
+	    INTEGER(columns)[i] = ncols(var_i);
+	}
+	else if (isNumeric(var_i)) {
+	    SET_VECTOR_ELT(variable, i, coerceVector(var_i, REALSXP));
+	    var_i = VECTOR_ELT(variable, i);
+	    LOGICAL(ordered)[i] = 0;
+	    INTEGER(nlevs)[i] = 0;
+	    INTEGER(columns)[i] = ncols(var_i);
+	}
+	else {
+	    LOGICAL(ordered)[i] = 0;
+	    INTEGER(nlevs)[i] = 0;
+	    INTEGER(columns)[i] = ncols(var_i);
+	}
+/*	else
+	    error(_("invalid variable type for '%s'"),
+	    CHAR(STRING_ELT(vnames, i))); */
+    }
+
+    /* If there is no intercept we look through the factor pattern */
+    /* matrix and adjust the code for the first factor found so that */
+    /* it will be coded by dummy variables rather than contrasts. */
+
+    if (!intrcept) {
+	for (j = 0; j < nterms; j++) {
+	    for (i = risponse; i < nVar; i++) {
+		if (INTEGER(nlevs)[i] > 1
+		    && INTEGER(factors)[i + j * nVar] > 0) {
+		    INTEGER(factors)[i + j * nVar] = 2;
+		    goto alldone;
+		}
+	    }
+	}
+    }
+ alldone:
+    ;
+
+    /* Compute the required contrast or dummy variable matrices. */
+    /* We set up a symbolic expression to evaluate these, substituting */
+    /* the required arguments at call time.  The calls have the following */
+    /* form: (contrast.type nlevs contrasts) */
+
+    PROTECT(contr1 = allocVector(VECSXP, nVar));
+    PROTECT(contr2 = allocVector(VECSXP, nVar));
+
+    PROTECT(expr = allocList(3));
```

Comment:
change this to `contr1 = Rf_lang3(R_NilValue, RNil_Value, RNil_Value)` to avoid the call to `SET_TYPEOF()`